File: data/code_pairs/cpp/1080.cpp
#include <iostream>
#include <memory>

class Base {
public:
    int x;
    Base() : x(0) {}
    virtual ~Base() = default;
    virtual int getval() const = 0;
};

class ChildAType : public Base {
public:
    ChildAType(int x) { this->x = x; }
    int getval() const override { return x; }
    static std::shared_ptr<ChildAType> alias_t(ChildAType& self) {
        return std::shared_ptr<ChildAType>(&self, [](ChildAType*){});
    }
    static std::shared_ptr<ChildAType> alias_c(const ChildAType& self) {
        return std::shared_ptr<ChildAType>(const_cast<ChildAType*>(&self), [](ChildAType*){});
    }
};

class ChildBType : public Base {
public:
    ChildBType(int x) { this->x = x; }
    int getval() const override { return x * 2; }
};

int main() {
    ChildAType a(10);
    ChildBType b(10);

    std::cout << "Testing getval() for ChildAType: ";
    std::cout << ((a.getval() == 10) ? "PASS" : "FAIL") << std::endl;

    std::cout << "Testing getval() for ChildBType: ";
    std::cout << ((b.getval() == 20) ? "PASS" : "FAIL") << std::endl;

    auto aliasT = ChildAType::alias_t(a);
    std::cout << "Testing alias_t for ChildAType: ";
    std::cout << ((aliasT->getval() == 10) ? "PASS" : "FAIL") << std::endl;

    auto aliasC = ChildAType::alias_c(a);
    std::cout << "Testing alias_c for ChildAType: ";
    std::cout << ((aliasC->getval() == 10) ? "PASS" : "FAIL") << std::endl;

    return 0;
}

File: data/code_pairs/fortran/1080.f90
MODULE class_pointers

  IMPLICIT NONE

  TYPE, ABSTRACT :: Base
    INTEGER :: x
  CONTAINS
    PROCEDURE(getval_template), DEFERRED :: getval
  END TYPE Base

  TYPE, EXTENDS(Base) :: ChildAType
  CONTAINS
    PROCEDURE :: getval => childa_getval
    PROCEDURE :: alias_t => childa_alias_t
    PROCEDURE :: alias_c => childa_alias_c
  END TYPE ChildAType

  TYPE, EXTENDS(Base) :: ChildBType
  CONTAINS
    PROCEDURE :: getval => childb_getval
  END TYPE ChildBType

  ABSTRACT INTERFACE
    FUNCTION getval_template(self) RESULT(val)
      IMPORT :: Base
      CLASS(Base), INTENT(IN) :: self
      INTEGER :: val
    END FUNCTION getval_template
  END INTERFACE

CONTAINS

  FUNCTION childa_getval(self) RESULT(val)
    CLASS(ChildAType), INTENT(IN) :: self
    INTEGER :: val
    val = self%x
  END FUNCTION childa_getval

  FUNCTION childa_alias_t(self) RESULT(p)
    CLASS(ChildAType), TARGET :: self
    TYPE(ChildAType), POINTER :: p
    p => self
  END FUNCTION childa_alias_t

  FUNCTION childa_alias_c(self) RESULT(p)
    CLASS(ChildAType), TARGET :: self
    CLASS(ChildAType), POINTER :: p
    p => self
  END FUNCTION childa_alias_c

  FUNCTION childb_getval(self) RESULT(val)
    CLASS(ChildBType), INTENT(IN) :: self
    INTEGER :: val
    val = self%x * 2
  END FUNCTION childb_getval

END MODULE class_pointers

PROGRAM test_class_pointers
  USE class_pointers
  IMPLICIT NONE

  CLASS(Base), POINTER :: p
  TYPE(ChildAType), POINTER :: a
  TYPE(ChildBType), POINTER :: b
  TYPE(ChildAType), POINTER :: aliasA
  CLASS(Base), POINTER :: aliasB

  ALLOCATE(ChildAType :: a)
  ALLOCATE(ChildBType :: b)
  a%x = 10
  b%x = 10

  p => a
  PRINT *, 'ChildAType getval:', p%getval()

  aliasA => childa_alias_t(a)
  PRINT *, 'ChildAType alias_t:', aliasA%getval()

  aliasA => childa_alias_c(a)
  PRINT *, 'ChildAType alias_c:', aliasA%getval()

  p => b
  PRINT *, 'ChildBType getval:', p%getval()

  DEALLOCATE(a, b)
END PROGRAM test_class_pointers